File: LC/[0256] Paint House.cpp
class Solution {
public:
    int minCost(vector<vector<int>>& costs) {
        int cost0=0, cost1=0, cost2=0;
        int N=costs.size();
        for(int ii=0; ii<N; ++ii){
            int cost0_tmp=cost0, cost1_tmp=cost1, cost2_tmp=cost2;

            if(ii==0){
                cost0=costs[ii][0];
                cost1=costs[ii][1];
                cost2=costs[ii][2];
            }else{
                cost0=min(cost1_tmp, cost2_tmp)+costs[ii][0];
                cost1=min(cost0_tmp, cost2_tmp)+costs[ii][1];
                cost2=min(cost0_tmp, cost1_tmp)+costs[ii][2];
            }
        }

        return min(min(cost0, cost1), cost2);
        
    }
};

File: LC/[0256] Paint House.java
class Solution {
    public int minCost(int[][] costs) {
        int N=costs.length;
        int rr=costs[0][0];   
        int gg=costs[0][1];      
        int bb=costs[0][2];     

        for(int ii=1; ii<N; ++ii){
            int tmpRR=rr;
            int tmpGG=gg;
            int tmpBB=bb;
            rr=Math.min(tmpGG, tmpBB) + costs[ii][0];
            gg=Math.min(tmpRR, tmpBB) + costs[ii][1];
            bb=Math.min(tmpRR, tmpGG) + costs[ii][2];
        }

        int[] res = {rr, gg, bb};
        return IntStream.of(res)//[java][min][array]
                     .min()
                     .orElseThrow(() -> new RuntimeException("Array is empty"));

    }
}
/*
- N=100
- current state is make a decision out of 3 options.  the current decision cannot be the same as prev
- similar to house robber.  two options (rob/noRob). the current decision cannot be same as prev
-> min ttl cost
*/